File: Sources/RelaySetup/BackupStore.swift
import Foundation

struct BackupRecord: Codable {
    let version: Int
    let client: ClientKind
    let createdAt: Date
    var originals: [FileSnapshot]
    var appliedHashes: [String: String]
}

final class BackupStore {
    let directory: URL
    private let encoder: JSONEncoder
    private let decoder = JSONDecoder()

    init(directory: URL? = nil) {
        if let directory {
            self.directory = directory
        } else if let override = ProcessInfo.processInfo.environment["RELAY_SETUP_SUPPORT_DIR"],
                  !override.isEmpty {
            self.directory = URL(fileURLWithPath: override, isDirectory: true)
        } else {
            let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            self.directory = support.appendingPathComponent("RelaySetup", isDirectory: true)
        }
        encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        decoder.dateDecodingStrategy = .iso8601
    }

    func record(for client: ClientKind) throws -> BackupRecord? {
        let url = recordURL(for: client)
        guard FileManager.default.fileExists(atPath: url.path) else { return nil }
        do {
            return try decoder.decode(BackupRecord.self, from: Data(contentsOf: url))
        } catch {
            throw RelaySetupError.file("备份记录损坏：\(error.localizedDescription)")
        }
    }

    func ensureBaseline(for client: ClientKind, paths: [URL]) throws {
        if var existing = try record(for: client) {
            let recordedPaths = Set(existing.originals.map(\.path))
            let additions = try paths
                .filter { !recordedPaths.contains($0.path) }
                .map(PrivateFileSystem.snapshot)
            if !additions.isEmpty {
                existing.originals.append(contentsOf: additions)
                try save(existing)
            }
            return
        }

        let originals = try paths.map(PrivateFileSystem.snapshot)
        let record = BackupRecord(
            version: 1,
            client: client,
            createdAt: Date(),
            originals: originals,
            appliedHashes: [:]
        )
        try save(record)
    }

    func markApplied(for client: ClientKind, paths: [URL]) throws {
        guard var record = try record(for: client) else { throw RelaySetupError.noBackup }
        record.appliedHashes = try Dictionary(uniqueKeysWithValues: paths.map {
            ($0.path, try PrivateFileSystem.hash(of: $0))
        })
        try save(record)
    }

    func status(for client: ClientKind) throws -> ConfigurationStatus {
        guard let record = try record(for: client) else { return .notConfigured }
        for (path, expectedHash) in record.appliedHashes {
            if try PrivateFileSystem.hash(of: URL(fileURLWithPath: path)) != expectedHash {
                return .changed
            }
        }
        return record.appliedHashes.isEmpty ? .notConfigured : .configured
    }

    func restore(client: ClientKind, force: Bool) throws {
        guard let record = try record(for: client) else { throw RelaySetupError.noBackup }
        if !force, try status(for: client) == .changed {
            throw RelaySetupError.restoreConflict
        }
        for snapshot in record.originals.reversed() {
            try PrivateFileSystem.restore(snapshot)
        }
        do {
            try FileManager.default.removeItem(at: recordURL(for: client))
        } catch {
            throw RelaySetupError.file(error.localizedDescription)
        }
    }

    private func save(_ record: BackupRecord) throws {
        do {
            let data = try encoder.encode(record)
            try PrivateFileSystem.write(data, to: recordURL(for: record.client))
        } catch let error as RelaySetupError {
            throw error
        } catch {
            throw RelaySetupError.file(error.localizedDescription)
        }
    }

    private func recordURL(for client: ClientKind) -> URL {
        directory.appendingPathComponent("\(client.rawValue)-backup.json")
    }
}

File: windows/RelayMate.Core/BackupStore.cs
using System.Text.Json;

namespace RelayMate.Core;

public sealed record BackupRecord(
    int Version,
    ClientKind Client,
    DateTimeOffset CreatedAt,
    List<FileSnapshot> Originals,
    Dictionary<string, string> AppliedHashes);

public sealed class BackupStore
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    public BackupStore(string directory)
    {
        Directory = directory;
    }

    public string Directory { get; }

    public BackupRecord? Record(ClientKind client)
    {
        var path = RecordPath(client);
        if (!File.Exists(path))
        {
            return null;
        }

        try
        {
            return JsonSerializer.Deserialize<BackupRecord>(File.ReadAllBytes(path), JsonOptions)
                ?? throw RelayMateException.File("备份记录为空。");
        }
        catch (JsonException exception)
        {
            throw RelayMateException.File($"备份记录损坏：{exception.Message}");
        }
    }

    public void EnsureBaseline(ClientKind client, IEnumerable<string> paths)
    {
        var existing = Record(client);
        if (existing is not null)
        {
            var recorded = existing.Originals.Select(item => item.Path).ToHashSet(StringComparer.OrdinalIgnoreCase);
            foreach (var path in paths.Where(path => !recorded.Contains(path)))
            {
                existing.Originals.Add(SecureFileSystem.Snapshot(path));
            }
            Save(existing);
            return;
        }

        Save(new BackupRecord(
            1,
            client,
            DateTimeOffset.UtcNow,
            paths.Select(SecureFileSystem.Snapshot).ToList(),
            new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)));
    }

    public void MarkApplied(ClientKind client, IEnumerable<string> paths)
    {
        var record = Record(client) ?? throw RelayMateException.NoBackup();
        record.AppliedHashes.Clear();
        foreach (var path in paths)
        {
            record.AppliedHashes[path] = SecureFileSystem.Hash(path);
        }
        Save(record);
    }

    public ConfigurationStatus Status(ClientKind client)
    {
        var record = Record(client);
        if (record is null || record.AppliedHashes.Count == 0)
        {
            return ConfigurationStatus.NotConfigured;
        }

        return record.AppliedHashes.Any(item => SecureFileSystem.Hash(item.Key) != item.Value)
            ? ConfigurationStatus.Changed
            : ConfigurationStatus.Configured;
    }

    public void Restore(ClientKind client, bool force)
    {
        var record = Record(client) ?? throw RelayMateException.NoBackup();
        if (!force && Status(client) == ConfigurationStatus.Changed)
        {
            throw RelayMateException.RestoreConflict();
        }

        foreach (var snapshot in record.Originals.AsEnumerable().Reverse())
        {
            SecureFileSystem.Restore(snapshot);
        }
        File.Delete(RecordPath(client));
    }

    private void Save(BackupRecord record)
    {
        SecureFileSystem.Write(JsonSerializer.SerializeToUtf8Bytes(record, JsonOptions), RecordPath(record.Client));
    }

    private string RecordPath(ClientKind client) =>
        Path.Combine(Directory, $"{client.ToString().ToLowerInvariant()}-backup.json");
}
